components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/persistence/cache/CacheBackedPolicyDAO.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2024, WSO2 LLC. (http://www.wso2.com).
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.entitlement.persistence.cache;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.entitlement.EntitlementException;
+import org.wso2.carbon.identity.entitlement.cache.PapPolicyCache;
+import org.wso2.carbon.identity.entitlement.cache.PapPolicyListCache;
+import org.wso2.carbon.identity.entitlement.cache.PdpPolicyCache;
+import org.wso2.carbon.identity.entitlement.cache.PdpPolicyListCache;
+import org.wso2.carbon.identity.entitlement.dto.PolicyDTO;
+import org.wso2.carbon.identity.entitlement.dto.PolicyStoreDTO;
+import org.wso2.carbon.identity.entitlement.persistence.dao.PolicyDAO;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class CacheBackedPolicyDAO extends PolicyDAO {
+
+    private static final Log LOG = LogFactory.getLog(CacheBackedPolicyDAO.class);
+    private final PapPolicyCache papPolicyCache = PapPolicyCache.getInstance();
+    private final PapPolicyListCache papPolicyListCache = PapPolicyListCache.getInstance();
+    private final PdpPolicyCache pdpPolicyCache = PdpPolicyCache.getInstance();
+    private final PdpPolicyListCache pdpPolicyListCache = PdpPolicyListCache.getInstance();
+
+    @Override
+    public void insertPolicy(PolicyDTO policy, int tenantId) throws EntitlementException {
+
+        super.insertPolicy(policy, tenantId);
+        papPolicyCache.addToCache(policy.getPolicyId(), policy, tenantId);
+        papPolicyListCache.clearCacheEntry(tenantId, tenantId);
```

Comment:
Is this method only add policy to PAP?
Wanted to validate whether we have to clear `pdpPolicyListCache` as well..

Reply:
when a new policy is added, it can only add to the PAP. found an issue with us still setting the IS_IN_PDP value based on the `isPromote` value. Normally it will not be true when a policy is first created, but since in DB impl we only use this method to add to PAP, have to set this to false each time, hence will fix that in a separate PR.